### PR TITLE
Add logic to handle communicator bindings in kubectl

### PR DESCRIPTION
**Pull request by mszostok**
## Description

Changes proposed in this pull request:

**Enhancements**
- Currently, we allowed to have only one executor binding on the list. 
  **Now executor manager is able to take the bindings list and calculate proper restrictions for a given channel.**
- E2E and unit test coverage to ensure no regression.
- Change `commands list` output:
    <table>
    <tr>
    <td> Before </td> <td> After </td>
    </tr>
    <tr>
    <td>

    ```yaml
    allowed verbs:
      - api-resources
      - api-versions                        
      - auth
      - cluster-info
      - describe
      - diff
      - explain
      - get
      - logs
      - top
    allowed resources:
      - configmaps
      - daemonsets
      - deployments
      - namespaces
      - nodes
      - pods
      - statefulsets
      - storageclasses
    allowed namespaces:
      include:
        - botkube
        - default
    ```

    </td>
    <td>

    ```yaml
    enabled:
      kubectl:
        kubectl-read-only:                          
          namespaces:
            include:
              - botkube
              - default
          enabled: true
          commands:
            verbs:
              - api-resources
              - api-versions
              - cluster-info
              - describe
              - diff
              - explain
              - get
              - logs
              - top
              - auth
            resources:
              - deployments
              - pods
              - namespaces
              - daemonsets
              - statefulsets
              - storageclasses
              - nodes
              - configmaps
          defaultNamespace: default
          restrictAccess: false
        kubectl-wait-cmd:
          namespaces:
            include:
              - botkube
              - default
          enabled: true
          commands:
            verbs:
              - wait
            resources: []
          restrictAccess: false`
    ```

    </td>
    </tr>
    </table>

**Bugs:**
-  On v0.12.4 `@BotKube  get po/botkube-56c8c5874b-f2bbm -n botkube` results in:
    ```
    Command not supported. Please run /botkubehelp to see supported commands.
    ```
    This was addressed on this PR. And now it's validated properly.
- On `main` the `-A` was returning all data even if not authorized `namespace.include: ["default"]`
   This was addressed on this PR. Now, the `-A` works only if you have binding with `namespace.include: ["all"]`
- On v0.12.4 `BotKube get -n botkube po` returns, 
    ```
    Command not supported. Please run /botkubehelp to see supported commands.
    ```
    On main returns:
    ```
    Sorry, the kubectl command is not authorized to work with '-n' resources on cluster 'dev'. Use 'commands list' to see all allowed resources.
    ```
    In this PR it was fixed to:
    ```
    Please specify the resource name after the verb, and all flags after the resource name. Format <verb> <resource> [flags]
    ```
    The better approach will be to remove all flags and find the resource name, however it's not simple. The proper solution should be to delegate this to `kubectl` which can do a proper parsing.

## Testing

The merger was fully tested with unit-tests 
- https://github.com/kubeshop/botkube/pull/672/files#diff-efacb45b078225d6d6058c5ca9152fb14cb291663d78dd64f3984bcae17d49dd
- https://github.com/kubeshop/botkube/pull/672/files#diff-189b9a00b5d538a5926c3c1543d0039ce151dd3e732dbefa346c103f4db44568

and e2e tests:
- It's tested as a part of the e2e tests. See new values: https://github.com/kubeshop/botkube/blob/c5b791b7ba3d0d0e07f12e2df9618af08658ee93/helm/botkube/e2e-test-values.yaml#L31-L48

It doesn't make sens to test the rest of the communication platform as the logic was not changed for them.

## Related issue(s)

Fixes https://github.com/kubeshop/botkube/issues/668

### CODE REVIEW

File: pkg/execute/kubectl/merger.go
Comment (mszostok) [inline pkg/execute/kubectl/merger.go:18]: We can create a wrapper to cache the calc responses. Example pseudo code:

```go
package kubectl

import (
	"fmt"

	"github.com/kubeshop/botkube/pkg/config"
)

type Merger interface {
	MergeForNamespace(includeBindings []string, forNamespace string) EnabledKubectl
	MergeAllEnabled(includeBindings []string) EnabledKubectl
	IsAtLeastOneEnabled() bool
	MergeAllEnabledVerbs(bindings []string) map[string]struct{}
}

type CachedMerger struct {
	underlying Merger
}

// NewMergerWithCache returns a new Merger instance with caching layer.
func NewMergerWithCache(merger Merger) *CachedMerger {
	return &CachedMerger{
		underlying: merger,
	}
}

func (m *CachedMerger) MergeAllEnabledVerbs(bindings []string) map[string]struct{} {
	const key = "MergeAllEnabledVerbs"
	m.mux[key].Lock()
	defer m.mux[key].UnLock()

	var entryKey = fmt.Sprintf("%s/%s", key, bindings)
	out, found := m.cache[entryKey]
	if !found {
		out = m.underlying.MergeAllEnabledVerbs(bindings)
		m.cache[entryKey] = out
	}

	return out
}

func (m *CachedMerger) MergeForNamespace(includeBindings []string, forNamespace string) EnabledKubectl {
 // TODO
}

func (m *CachedMerger) MergeAllEnabled(includeBindings []string) EnabledKubectl {
 // TODO
}

func (m *CachedMerger) IsAtLeastOneEnabled() bool {
 // TODO
}

```

We can also use some caching Go lib, e.g. https://github.com/allegro/bigcache

File: pkg/bot/slack.go
Comment (pkosiec) [inline pkg/bot/slack.go:48]: Hmm 🤔 I tried to test it and:

```
      channels:
        'default':
          # -- Slack channel name without '#' prefix where you have added BotKube and want to receive notifications in.
          name: 'SLACK_CHANNEL'
          bindings:
            executors:
              - kubectl-read-only
              - kubectl-read-only2
            sources:
              - k8s-events`
```

```
executors:
  'kubectl-read-only2':
    ## Kubectl executor configuration.
    kubectl:
      namespaces:
        # -- List of allowed Kubernetes Namespaces for command execution. The "all" keyword, allows all the Namespaces.
        include:
          - all
        # -- List of ignored Kubernetes Namespace. It can be used only when `include: ["all"]`.
        # It can contain a wildcard (*) that expands to zero or more arbitrary characters
        # Example: `ignore: ["x", "y", "secret-ns-*"]`
        ignore: []
      # -- If true, enables `kubectl` commands execution.
      enabled: true
      ## List of allowed `kubectl` commands.
      commands:
        # -- Configures which `kubectl` methods are allowed.
        verbs: ["get", "logs", "top", "auth"]
        # -- Configures which K8s resource are allowed.
        resources: ["deployments", "pods"]
      # -- Configures the default Namespace for executing BotKube `kubectl` commands. If not set, uses the 'default'.
      defaultNamespace: default
      # -- If true, enables commands execution from configured channel only.
      restrictAccess: true
  'kubectl-read-only':
    ## Kubectl executor configuration.
    kubectl:
      namespaces:
        # -- List of allowed Kubernetes Namespaces for command execution. The "all" keyword, allows all the Namespaces.
        include:
          - all
        # -- List of ignored Kubernetes Namespace. It can be used only when `include: ["all"]`.
        # It can contain a wildcard (*) that expands to zero or more arbitrary characters
        # Example: `ignore: ["x", "y", "secret-ns-*"]`
        ignore: []
      # -- If true, enables `kubectl` commands execution.
      enabled: true
      ## List of allowed `kubectl` commands.
      commands:
        # -- Configures which `kubectl` methods are allowed.
        verbs: ["api-resources", "api-versions", "cluster-info", "describe", "diff", "explain"]
        # -- Configures which K8s resource are allowed.
        resources: [ "namespaces", "daemonsets", "statefulsets", "storageclasses", "nodes", "configmaps"]
      # -- Configures the default Namespace for executing BotKube `kubectl` commands. If not set, uses the 'default'.
      defaultNamespace: default
      # -- If true, enables commands execution from configured channel only.
      restrictAccess: true
```

Screenshot 2022-08-04 at 13 22 10

Screenshot 2022-08-04 at 13 22 18


Looks like the `kubectl-read-only2` binding doesn't work, even if I have that in `commands list`:
Screenshot 2022-08-04 at 13 22 59

Do you know what could be the problem?

**Review by pkosiec — COMMENTED** [review]
Looks great 🚀 I have one open question when it comes to the behavior. Not sure if that's a bug or I don't understand it (probably the latter). But overall LGTM once we sync f2f about it ✅

File: pkg/bot/discord.go
Comment (mszostok) [reply pkg/bot/discord.go:48]: yep, it was for testing, I didn't want to duplicate your work 👍 so this is a small step for handling binding where still you have opt to have only one channel.

File: pkg/bot/slack.go
Comment (pkosiec) [reply pkg/bot/slack.go:48]: Alright, that was issue on my side, thanks for the sync 😄

**Review by pkosiec — APPROVED** [review]
Works well 👌 Please see the comments before merge.

File: pkg/execute/kubectl/merger.go
Comment (ezodude) [inline pkg/execute/kubectl/merger.go:62]: Having had a cursory look at `MergeAllEnabledVerbs` is used.

To get the semantics right, what if it simply returns a Set? (`map[string]bool`) ?

As a freebie, you just get the value and wont't need `_, ok := ...`

File: pkg/execute/kubectl.go
Comment (ezodude) [inline pkg/execute/kubectl.go:57]: Slight nit pick.

What if the merger simply told you if it includes a verb? e.g. `e.merger.includesVerb(args[0]])`

**Review by ezodude — COMMENTED** [review]
That is one big PR! 😅

I understand the premise. And, left a few small comments mainly regarding style and less indirection.

Will run the code before I approve. 🙂

File: pkg/execute/executor.go
Comment (mszostok) [reply pkg/execute/executor.go:390]: I will just replace the `enabled` with `Enabled executors` in that way the indent will be more predictable 👍

File: pkg/execute/executor.go
Comment (mszostok) [reply pkg/execute/executor.go:397]: I used that previously because in default setting it is using 4 spaces, and that was not in sync with the data that was printed "manually" with:
```go
	const itemSeparator = "\n  - "
	items := strings.Join(keys, itemSeparator)
```

Now we use only `yaml.Marshal` but I still think that 2 spaces are ok and not 4 are a bit too wide in the communicator window.

File: pkg/execute/kubectl.go
Comment (mszostok) [reply pkg/execute/kubectl.go:251]: wow, that's neat! 
```go
func Test(t *testing.T) {

	resource, _, _ := strings.Cut("po/test", "/")
	fmt.Println(resource)

	resource, _, _ = strings.Cut("po/test/foo", "/")
	fmt.Println(resource)

	resource, _, _ = strings.Cut("foo", "/")
	fmt.Println(resource)

	resource, _, _ = strings.Cut("/foo", "/")
	fmt.Println(resource)
}
```

so I just simplified to 
```go
func (e *Kubectl) getResourceName(args []string) string {
	if len(args) < 2 {
		return ""
	}
	resource, _, _ := strings.Cut(args[1], "/")
	return resource
}
```

File: pkg/execute/executor.go
Comment (mszostok) [reply pkg/execute/executor.go:141]: >  I think the latter option would be more consistent

Make sens, so feel free to replace that on you PR 👍

File: pkg/execute/executor_test.go
Comment (mszostok) [reply pkg/execute/executor_test.go:145]: The `fixtureExecutorsConfig` is a short for `fixtureExecutorsConfig`. We had this approach and other methods used in testing has this `fix` prefix. Let me know if it's ok for you 👍

File: pkg/execute/kubectl.go
Comment (mszostok) [reply pkg/execute/kubectl.go:57]: IMO it would be good to have that separated. For now, the boundaries are quite easy to describe. Merger takes care of any sort of merging strategies. Checker, on the other hand, it's used to assert some expectation. I'm not sure if it's worth to mix this functionality even if it's simple. I tried to avoid having a one component that will do all stuff, instead wanted to provide smaller components where the logic is simple enough to be well tested and read.

**Comment by mszostok** [comment]
Hi @ezodude 
From the output that you posted, I can see that you are not using the image from this PR. Please use:
```
--set image.repository=kubeshop/pr/botkube \
--set image.tag=672-PR \
```
In the e2e tests, you can see that we test that scenario. I specify multiple executors:
https://github.com/kubeshop/botkube/blob/d12b76ac1a1ee85c3ed2125d6a6a7dd6e394e34b/helm/botkube/e2e-test-values.yaml#L34-L70

and assert that only those enabled would be displayed:
https://github.com/kubeshop/botkube/blob/d12b76ac1a1ee85c3ed2125d6a6a7dd6e394e34b/test/e2e/slack_test.go#L124-L175

The new format was also mentioned in the PR description. 

It has to be changed as a given list of resources and verbs need to be associated with a given set of namespaces. I could try to come up with some merging strategy. However, for now, IMO it's good to print the config as it is so user don't need to be familiar with different syntaxes. In the long term, it would be nice to revisit all BotKube's commands and come up with some consistent approach (better UX). In this way, configuration would be nicely presented (with some field/modals?) to the user, but he will still have an option to get plain config format e.g. using `raw` flag.

File: pkg/execute/kubectl/merger.go
Comment (mszostok) [reply pkg/execute/kubectl/merger.go:62]: I learned that a long time ago. I'm using the `map[string]struct{}` not because it's give us [allocation improvements](https://dave.cheney.net/2014/03/25/the-empty-struct) but it gives more explicit API 



>I use and will continue to use map[string]struct{}, because when someone reads that, they know, beyond a shadow of a doubt, that existence is the only thing that matters and there is no place in the code, no matter how far this map travels, that will ever make a decision based on anything else. When I used map[string]bool, the person reading the code does NOT know that. You can assert it in a comment all you like, but the reader must still consult the code to verify that it can't go anywhere where some other programmer (who may even be you at some other point in time) didn't decide to make the bool parameter mean something. 

So when I see `map[string]bool` I start thinking if there is a functionality that is able to enable/disable it, if yes then on what condition? 
On the other hand, using `[]string` give us O(n) instead of `O(1)`. 

I'm open to discuss it further and establish an approach that would be suitable for both of us 👍

File: pkg/execute/executor.go
Comment (pkosiec) [reply pkg/execute/executor.go:397]: Alright, but it would be great to have consistency. But keep it as it is, we'll definitely improve configs printed later 👍

File: pkg/execute/kubectl.go
Comment (pkosiec) [reply pkg/execute/kubectl.go]: Right, I forgot about this: `// If sep does not appear in s, cut returns s, "", false.` 😄

**Review by pkosiec — APPROVED** [review]
LGTM

**Comment by mszostok** [comment]
Merging this PR as it's blocking https://github.com/kubeshop/botkube/pull/670

File: pkg/execute/kubectl/merger.go
Comment (ezodude) [reply pkg/execute/kubectl/merger.go:62]: All good. 👍

File: pkg/execute/kubectl.go
Comment (ezodude) [reply pkg/execute/kubectl.go:57]: I see what you're getting at - I just felt there's a bit of jumping around.

Ensuring the components are decoupled trumps everything. Here, I felt a generic (`includes...`)[tell don't ask](https://martinfowler.com/bliki/TellDontAsk.html) might help.

It definitely should not delay merging this PR - so all good.

**Pull request merged**
